Test/paulmck/rcu.hpp

```diff
@@ -3,45 +3,26 @@
 #include <cstddef>
 #include <memory>
 #include <mutex>
+#include <tuple>
 #include <utility>
 
 // Derived-type approach.  All RCU-protected data structures using this
 // approach must derive from std::rcu_obj_base, which in turn derives
 // from std::rcu_head.  No idea what happens in case of multiple inheritance.
 
 namespace std {
-    template<typename T, typename D = default_delete<T>, bool E = is_empty<D>::value>
-    class rcu_obj_base: private rcu_head {
-        D deleter;
+    template<typename T, typename D = default_delete<T>>
+    class rcu_obj_base: private rcu_head, private std::tuple<D> {
```

Comment:
Nifty. I've never seen this pattern before and am thus mildly skeptical of it; but I don't immediately see anything wrong with it. It's also an improvement over my hack (on old line 44) of default-constructing a deleter every time we need one — that hack was definitely bad.

Does the Standard (or at least any major vendor) actually guarantee that `is_empty_v<tuple<E>>` whenever `is_empty_v<E>`?